### ISSUE TITLE
Generate and store vertex normals in .bsp for smooth (phong) shading - possible ?

### ISSUE THREAD

**Issue by motorsep** [qbism/q2tools-220#17]
Since Q2RTX doesn't use lightmaps, currently it's not possible to have smoothed (phong) shading in Q2 maps.

Could you please add such functionality, if possible, to 4bsp? It would need to generate and store vertex normals somewhere (in the vertex lump?) in .bsp file.

Q2RTX dev will add support for that, if he has something to work with. Thanks!

**Comment by Paril** [comment]
I'm thinking the approach we'll be taking with another tool is to simply use an unused surface flag (0x400000 maybe?) to represent "this surface was compiled with phong", which would allow RTX to handle as appropriate to reconstruct the normals.
The other option is to have qbism and other tools export a normal map, similar to a deluxemap, which RTX could use instead of reconstructing the normals.

**Comment by qbism** [comment]
How does Berserker Q2 do it?  From overview.doc:

To improve the quality of the lighting of the map and its parts (brush models), inter-vertex smoothing is applied taking into account the mutual inclination of the planes. If the angle of intersection of two planes is close to 180 degrees (+/- angle, which can be changed), then we perform the smoothing of the TBN-basis of vectors (Tangent, Binormal, Normal) for them. Visually, it looks like this:
![berserker-phong](https://user-images.githubusercontent.com/1802223/130880482-c8ae26f5-c5ec-4a28-9642-a5b9cda02049.png)

**Comment by motorsep** [comment]
@qbism I think it's all done in the lighting stage, but I am no pro in this subject 😊 
Paril would probably have more input about that.

**Comment by Paril** [comment]
Yeah it's not really difficult implementation wise, phong has been a
feature of lightmap pers forever, but the hard part is delivering the raw
normals/tangents to the engine which rtx will need.

The main rtx dev (AlexP) , ericw and I discussed options (doing it at load
time, surface flags, external normal map..) and we think BSPX is our best
option. It's basically an extension footer that allows an existing bsp to
add additional lumps without breaking the format. This would allow q2bsp
and qbsp to do the phonging in the lightmap as usual, and the additional
lump would be picked up by rtx for the additional data, without the maps
not working in other engines as normal.
